File: 106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private Dictionary<int, int> inMap = new Dictionary<int, int>();
    private int[] inorder;
    private int[] postorder;
    
    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        this.inorder = inorder;
        this.postorder = postorder;
        
        for (int i = 0; i < inorder.Length; i++) {
            inMap[inorder[i]] = i;
        }
        
        return BuildTreeHelper(0, inorder.Length - 1, 0, postorder.Length - 1);
    }
    
    private TreeNode BuildTreeHelper(int inStart, int inEnd, int postStart, int postEnd) {
        if (inStart > inEnd) return null;
        
        TreeNode root = new TreeNode(postorder[postEnd]);
        int rootIndex = inMap[root.val];
        int leftSubtreeSize = rootIndex - inStart;
        
        root.left = BuildTreeHelper(inStart, rootIndex - 1,
                                  postStart, postStart + leftSubtreeSize - 1);
        root.right = BuildTreeHelper(rootIndex + 1, inEnd,
                                   postStart + leftSubtreeSize, postEnd - 1);
        
        return root;
    }
}

File: 106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    unordered_map<int, int> inMap;
    vector<int> inorder;
    vector<int> postorder;
    
    TreeNode* buildTreeHelper(int inStart, int inEnd, int postStart, int postEnd) {
        if (inStart > inEnd) return nullptr;
        
        TreeNode* root = new TreeNode(postorder[postEnd]);
        int rootIndex = inMap[root->val];
        int leftSubtreeSize = rootIndex - inStart;
        
        root->left = buildTreeHelper(inStart, rootIndex - 1,
                                   postStart, postStart + leftSubtreeSize - 1);
        root->right = buildTreeHelper(rootIndex + 1, inEnd,
                                    postStart + leftSubtreeSize, postEnd - 1);
        
        return root;
    }
    
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        this->inorder = inorder;
        this->postorder = postorder;
        
        for (int i = 0; i < inorder.size(); i++) {
            inMap[inorder[i]] = i;
        }
        
        return buildTreeHelper(0, inorder.size() - 1, 0, postorder.size() - 1);
    }
};